File: games/wavelength/backend/app/schemas/ws_events.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.room import RoomOut

# ---- Server -> Client ----


class RoomStateEvent(BaseModel):
    type: Literal["room_state"] = "room_state"
    room: RoomOut


class PlayerConnectionChangedEvent(BaseModel):
    type: Literal["player_connection_changed"] = "player_connection_changed"
    player_id: str
    connected: bool


class ErrorEvent(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


class PongEvent(BaseModel):
    type: Literal["pong"] = "pong"


class KickedEvent(BaseModel):
    type: Literal["kicked"] = "kicked"


class PsychicTargetEvent(BaseModel):
    """The target position — sent ONLY to the current Psychic's socket.
    Never broadcast. Sent at round start and on reconnect.
    """

    type: Literal["psychic_target"] = "psychic_target"
    target_position: float


class ClueGivenEvent(BaseModel):
    """Public: Psychic has given their clue (optional typed text)."""

    type: Literal["clue_given"] = "clue_given"
    clue_text: str
    psychic_id: str


class PlayerGuessedEvent(BaseModel):
    """Public: count-only update when a guesser locks in."""

    type: Literal["player_guessed"] = "player_guessed"
    player_id: str
    guessed_count: int


class RevealedEvent(BaseModel):
    """Public: target revealed + every guesser's position + round points."""

    type: Literal["revealed"] = "revealed"
    target_position: float
    guesses: dict[str, float]
    points_awarded: dict[str, int]


class GameOverEvent(BaseModel):
    """Public: game over with final scores."""

    type: Literal["game_over"] = "game_over"
    scores: dict[str, int]


# ---- Client -> Server ----


class PingCommand(BaseModel):
    type: Literal["ping"] = "ping"


class SetReadyCommand(BaseModel):
    type: Literal["set_ready"] = "set_ready"
    ready: bool


class UpdateProfileCommand(BaseModel):
    type: Literal["update_profile"] = "update_profile"
    display_name: str | None = None
    avatar: str | None = None


class KickPlayerCommand(BaseModel):
    type: Literal["kick_player"] = "kick_player"
    target_player_id: str


class LeaveRoomCommand(BaseModel):
    type: Literal["leave_room"] = "leave_room"


class StartGameCommand(BaseModel):
    type: Literal["start_game"] = "start_game"


class GiveClueCommand(BaseModel):
    type: Literal["give_clue"] = "give_clue"
    clue_text: str = ""


class SubmitGuessCommand(BaseModel):
    type: Literal["submit_guess"] = "submit_guess"
    position: float


class RevealCommand(BaseModel):
    type: Literal["reveal"] = "reveal"


class NextRoundCommand(BaseModel):
    type: Literal["next_round"] = "next_round"

File: games/wavelength/frontend/types/ws-events.ts
import type { Room } from "./room";

// ---- Server -> Client ----

export interface RoomStateEvent {
  type: "room_state";
  room: Room;
}

export interface PlayerConnectionChangedEvent {
  type: "player_connection_changed";
  player_id: string;
  connected: boolean;
}

export interface ErrorEvent {
  type: "error";
  code: string;
  message: string;
}

export interface PongEvent {
  type: "pong";
}

export interface KickedEvent {
  type: "kicked";
}

export interface PsychicTargetEvent {
  type: "psychic_target";
  /** The hidden target position (0–100). Sent ONLY to the current Psychic. */
  target_position: number;
}

export interface ClueGivenEvent {
  type: "clue_given";
  clue_text: string;
  psychic_id: string;
}

export interface PlayerGuessedEvent {
  type: "player_guessed";
  player_id: string;
  guessed_count: number;
}

export interface RevealedEvent {
  type: "revealed";
  target_position: number;
  guesses: Record<string, number>;
  points_awarded: Record<string, number>;
}

export interface GameOverEvent {
  type: "game_over";
  scores: Record<string, number>;
}

export type ServerEvent =
  | RoomStateEvent
  | PlayerConnectionChangedEvent
  | ErrorEvent
  | PongEvent
  | KickedEvent
  | PsychicTargetEvent
  | ClueGivenEvent
  | PlayerGuessedEvent
  | RevealedEvent
  | GameOverEvent;

// ---- Client -> Server ----

export interface PingCommand {
  type: "ping";
}

export interface SetReadyCommand {
  type: "set_ready";
  ready: boolean;
}

export interface UpdateProfileCommand {
  type: "update_profile";
  display_name?: string;
  avatar?: string;
}

export interface KickPlayerCommand {
  type: "kick_player";
  target_player_id: string;
}

export interface LeaveRoomCommand {
  type: "leave_room";
}

export interface StartGameCommand {
  type: "start_game";
}

export interface GiveClueCommand {
  type: "give_clue";
  clue_text?: string;
}

export interface SubmitGuessCommand {
  type: "submit_guess";
  position: number;
}

export interface RevealCommand {
  type: "reveal";
}

export interface NextRoundCommand {
  type: "next_round";
}

export type ClientCommand =
  | PingCommand
  | SetReadyCommand
  | UpdateProfileCommand
  | KickPlayerCommand
  | LeaveRoomCommand
  | StartGameCommand
  | GiveClueCommand
  | SubmitGuessCommand
  | RevealCommand
  | NextRoundCommand;
